File: iosApp/iosApp/Modules/Networks/NetworksWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol NetworksWireframeFactory {
    func makeWireframe(_ parent: UIViewController) -> NetworksWireframe
}

final class DefaultNetworksWireframeFactory {
    private let alertWireframeFactory: AlertWireframeFactory
    private let networkSettingsWireframeFactory: NetworkSettingsWireframeFactory
    private let networksService: NetworksService

    init(
        alertWireframeFactory: AlertWireframeFactory,
        networkSettingsWireframeFactory: NetworkSettingsWireframeFactory,
        networksService: NetworksService
    ) {
        self.alertWireframeFactory = alertWireframeFactory
        self.networkSettingsWireframeFactory = networkSettingsWireframeFactory
        self.networksService = networksService
    }
}

extension DefaultNetworksWireframeFactory: NetworksWireframeFactory {

    func makeWireframe(_ parent: UIViewController) -> NetworksWireframe {
        DefaultNetworksWireframe(
            parent: parent,
            networksService: networksService,
            alertWireframeFactory: alertWireframeFactory,
            networkSettingsWireframeFactory: networkSettingsWireframeFactory
        )
    }
}

// MARK: - NetworksWireframeFactoryAssembler

final class NetworksWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> NetworksWireframeFactory in
            DefaultNetworksWireframeFactory(
                alertWireframeFactory: resolver.resolve(),
                networkSettingsWireframeFactory: resolver.resolve(),
                networksService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/networks/NetworksWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.networks

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.networks.NetworksWireframe
import smartadapter.internal.extension.name

interface NetworksWireframeFactory {
    fun make(parent: Fragment?): NetworksWireframe
}

class DefaultNetworksWireframeFactory(
    private val networksService: NetworksService,
): NetworksWireframeFactory {

    override fun make(parent: Fragment?): NetworksWireframe = DefaultNetworksWireframe(
        parent?.let { WeakRef(it) },
        networksService,
    )
}

class NetworksWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(NetworksWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultNetworksWireframeFactory(
                it.resolve(NetworksService::class.name),
            )
        }
    }
}